File: Leetcode/数据结构/树的子结构.cpp
//区别另一个树的子树这道题，子树那道题需要调用Same函数
//和这题不同
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSubStructure(TreeNode* A, TreeNode* B) {
        if(!A || !B) return false;
        return helper(A,B) || isSubStructure(A->left,B) || isSubStructure(A->right,B);
    }

    bool helper(TreeNode* A,TreeNode* B){
        if(!B) return true;
        if(!A) return false;
        if(A->val!=B->val) return false;
        return helper(A->left,B->left) && helper(A->right,B->right);
    }
};

File: Leetcode/数据结构/树的子结构.py
class Solution:
    def check(self, s, t): # 这里 t 可能为空
        if not t:
            return True
        if not s:
            return False
        
        return s.val == t.val and self.check(s.left, t.left) and self.check(s.right, t.right)

    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        if not A or not B: return False
        return self.check(A, B) or self.isSubStructure(A.left, B) or self.isSubStructure(A.right, B)
